File: homework/thinhbka/lesson5-graph/207. Course Schedule.py
class Solution:
    def dfs(self,node):
        self.visited[node] = True
        self.is_visiting[node] = True
        for e in self.edges[node]:
            if self.visited[e] == False:
                self.dfs(e)
            else:
                if self.is_visiting[e] == True:
                    self.can_finish = False
                    
        self.is_visiting[node] = False
    
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        self.edges  =[[] for _ in range(numCourses)]
        self.can_finish = True
        for pair in prerequisites:
            self.edges[pair[0]].append(pair[1])
        self.visited = [False for _ in range(numCourses)]
        self.is_visiting = [False for _ in range(numCourses)]
        for i in range(numCourses):
            if self.visited[i] is False:
                self.dfs(i)
        return self.can_finish

File: homework/tai_nguyen/lesson7.2/207_course_schedule.ts
function canFinish(numCourses: number, prerequisites: number[][]): boolean {
  const edges: number[][] = new Array(numCourses).fill(true).map(() => []);
  prerequisites.forEach(([j, i]) => {
    edges[i].push(j);
  });

  const visited = new Array(numCourses).fill(false);
  const isOnStack = new Array(numCourses).fill(false);

  function dfs(i: number) {
    visited[i] = true;
    isOnStack[i] = true;

    for (const j of edges[i]) {
      if (isOnStack[j]) return false;
      if (!visited[j]) {
        const flag = dfs(j);
        if (!flag) return false;
      }
    }

    isOnStack[i] = false;
    return true;
  }

  for (let i = 0; i < numCourses; i++) {
    if (!visited[i] && !dfs(i)) {
      return false;
    }
  }

  return true;
}
